Code/Framework/AzToolsFramework/AzToolsFramework/Prefab/Undo/PrefabUndoEntityOverrides.cpp

```diff
@@ -0,0 +1,172 @@
+
+                // Get the before-state entity DOM inside the source template DOM of the top instance.
+                const TemplateId topTemplateId = link->get().GetSourceTemplateId();
+                const PrefabDom& topTemplateDom = m_prefabSystemComponentInterface->FindTemplateDom(topTemplateId);
+
+                PrefabDom overridePatchesWithSubpaths; // subpath is path pointing from entity to component value
+                {
+                    // This scope is added to limit their usage and ensure DOM is not modified when it is being used.
+                    // DOM value pointers can't be relied upon if the original DOM gets modified after pointer creation.
+                    const PrefabDomValue* entityDomInTopTemplate = entityDomPathFromTopInstance.Get(topTemplateDom);
+
+                    if (entityDomInTopTemplate)
+                    {
+                        m_instanceToTemplateInterface->GeneratePatch(
+                            overridePatchesWithSubpaths, *entityDomInTopTemplate, entityDomAfterUpdate);
+                    }
+                    else if (auto overrideType = m_prefabOverridePublicInterface->GetOverrideType(entity->GetId());
+                        overrideType.has_value() && overrideType == OverrideType::AddEntity)
+                    {
+                        // Check if the entity itself is added as an override already.
+                        // In this case, additional override patches on this entity would be merged into the entity DOM value that is
+                        // stored in the existing add-entity override patch.
+
+                        overridePatchesWithSubpaths.SetArray();
+                        // Note that for add-entity patch, there is no subpath that points from entity to component value.
+                        // Entity path passing to the util function is empty such that the generated JSON patch won't contain any path.
+                        // The path inside JSON patch will be set correctly below when generating the override subtree.
+                        PrefabUndoUtils::AppendAddEntityPatch(overridePatchesWithSubpaths, entityDomAfterUpdate, "");
+                    }
+                    else
+                    {
+                        AZ_Warning("Prefab", false, "PrefabUndoEntityOverrides::Capture - "
+                            "Entity ('%s') is not present in template for override editing. Skipping it.", entity->GetName().c_str());
+                        continue;
+                    }
+                }
+
+                // Remove the subtree and cache the subtree for undo.
+                AZ::Dom::Path entityDomPath(entityPathFromTopInstance.c_str());
+                m_subtreeStates[entityDomPath] = AZStd::move(link->get().RemoveOverrides(entityDomPath));
+
+                // Redo - Add the override patches to the tree.
+                link->get().AddOverrides(entityDomPath, overridePatchesWithSubpaths);
```

Comment:
Rather than making the link class append the path to each patch, we can do that during the addition of patch itself in line 111. This will also make the public function in the link class more generic and reusable for other use cases. I recommend that since its a good practice to keep the base api classes as minimalistic and generic as possible and leave custom logic to classes like this. Please update function comments too if you do that.

Reply:
Good callout! The previous workflow was for GenerateSubtree. After changing to AddOverrides(patches), it can be simplified a lot.